state/factory/factory.go

```diff
@@ -408,16 +408,35 @@ func (sf *factory) WorkingSet(ctx context.Context) (protocol.StateManager, error
 	return sf.newWorkingSet(ctx, sf.currentChainHeight+1)
 }
 
-func (sf *factory) WorkingSetAtHeight(ctx context.Context, height uint64) (protocol.StateManager, error) {
+func (sf *factory) WorkingSetAtHeight(ctx context.Context, height uint64, preacts ...*action.SealedEnvelope) (protocol.StateManager, error) {
 	if !sf.saveHistory {
 		return nil, ErrNoArchiveData
 	}
 	sf.mutex.Lock()
-	defer sf.mutex.Unlock()
 	if height > sf.currentChainHeight {
+		sf.mutex.Unlock()
 		return nil, errors.Errorf("query height %d is higher than tip height %d", height, sf.currentChainHeight)
 	}
-	return sf.newWorkingSetAtHeight(ctx, height)
+	var (
+		ws  *workingSet
+		err error
+	)
+	if len(preacts) == 0 {
+		ws, err = sf.newWorkingSet(ctx, height)
+		sf.mutex.Unlock()
+		return ws, err
+	}
+	// prepare workingset at height-1, and run acts
+	ws, err = sf.newWorkingSet(ctx, height-1)
```

Comment:
It's more clear to use height and pass height-1 at caller

Reply:
it is to get to a particular state, say after action 3 at block 1010, so need the 3 steps:
```
ws = sf.newWorkingSet(ctx, 1009)
ws.height++
ws.runActions(act 0, 1, 2)
```